engine/access/ingestion/engine.go

```diff
@@ -187,6 +187,34 @@ func New(
 	return e, nil
 }
 
+func (e *Engine) Start(parent irrecoverable.SignalerContext) {
+	rootBlock, err := e.state.Params().Root()
+	if err != nil {
+		parent.Throw(fmt.Errorf("failed to get root block: %w", err))
+	}
+
+	rootSnapshot := e.state.AtBlockID(rootBlock.ID())
+
+	isSporkRootSnapshot, err := protocol.IsSporkRootSnapshot(rootSnapshot)
+	if err != nil {
+		parent.Throw(fmt.Errorf("could not check if root snapshot is a spork root snapshot: %w", err))
+	}
+
+	// This is useful for dynamically bootstrapped access node, they will request missing collections. In order to ensure all txs
+	// from the missing collections can be verified, we must ensure they are referencing to known blocks.
+	// That's why we set the full block height to be rootHeight + TransactionExpiry, so that we only request missing collections
+	// in blocks above that height.
+	if !isSporkRootSnapshot {
+		firstFullHeight := rootBlock.Height + flow.DefaultTransactionExpiry
+		err := e.blocks.InsertLastFullBlockHeight(firstFullHeight)
```

Comment:
This call will be called multiple times right?

Usually insert operation will fail when a key already exists, which will cause it to crash.

How does it work here? Does it return nil for the second operation? We'd better comment about the behavior or describe the behavior in function name, such as `InsertLastFullBlockHeightIfNotExist`

If we are renaming to `InsertLastFullBlockHeightIfNotExist`, then we should handle the `storage.ErrAlreadyExists` error with `SkipDuplicates`.

Reply:
This will be called multiple times if the engine crashes, but will only insert the last full block height once if it doesn't exist. [SkipDuplicates](https://github.com/onflow/flow-go/blob/khalil/1761-an-finalized-block-request/storage/badger/operation/heights.go#L44) is already handling the `storage.ErrAlreadyExists ` for us. Functionality is already documented in the [interface](https://github.com/onflow/flow-go/blob/khalil/1761-an-finalized-block-request/storage/blocks.go#L36)  . I renamed the func for more readability https://github.com/onflow/flow-go/pull/2974/commits/0e04259bcc147991a8f9cc7db6f10a98496acf25.